File: scripts/sync-leaderboard.js
/**
 * 排行榜数据同步脚本
 * 每天23:00自动从参考网站抓取最新数据并更新
 */

const https = require('https');
const fs = require('fs');
const path = require('path');

// 配置
const CONFIG = {
    sourceUrl: 'https://model.aibase.com/zh/leaderboard/text-generation',
    targetFile: path.join(__dirname, '../js/model-center.js'),
    indexFile: path.join(__dirname, '../index.html'),
    logFile: path.join(__dirname, '../logs/sync-log.txt')
};

// 日志函数
function log(message) {
    const timestamp = new Date().toISOString();
    const logMessage = `[${timestamp}] ${message}\n`;
    console.log(logMessage.trim());
    
    // 确保logs目录存在
    const logsDir = path.dirname(CONFIG.logFile);
    if (!fs.existsSync(logsDir)) {
        fs.mkdirSync(logsDir, { recursive: true });
    }
    
    fs.appendFileSync(CONFIG.logFile, logMessage);
}

// 抓取网页内容
function fetchWebContent(url) {
    return new Promise((resolve, reject) => {
        https.get(url, (res) => {
            let data = '';
            
            res.on('data', (chunk) => {
                data += chunk;
            });
            
            res.on('end', () => {
                resolve(data);
            });
        }).on('error', (err) => {
            reject(err);
        });
    });
}

// 解析HTML，提取排行榜数据
function parseLeaderboardData(html) {
    const models = [];
    
    // 使用正则表达式提取数据
    // 这里需要根据实际HTML结构调整正则表达式
    const modelPattern = /#(\d+)\s*\[([^\]]+)\]\([^\)]+\)\s*([^\s]+)\s*([^\s]+)\s*(\d+)/g;
    
    let match;
    while ((match = modelPattern.exec(html)) !== null) {
        models.push({
            rank: parseInt(match[1]),
            model: match[2],
            provider: match[3],
            license: match[4],
            score: parseInt(match[5])
        });
    }
    
    // 如果正则提取失败，尝试备用方法
    if (models.length === 0) {
        log('使用备用解析方法...');
        // 这里可以添加更复杂的HTML解析逻辑
        // 或者使用cheerio等库
    }
    
    return models;
}

// 更新model-center.js文件
function updateModelData(models) {
    try {
        let content = fs.readFileSync(CONFIG.targetFile, 'utf8');
        
        // 构建新的数据数组
        const newDataStr = JSON.stringify(models, null, 8);
        
        // 替换leaderboard数据
        const pattern = /leaderboard:\s*\[[\s\S]*?\]/;
        const replacement = `leaderboard: ${newDataStr}`;
        
        content = content.replace(pattern, replacement);
        
        // 写入文件
        fs.writeFileSync(CONFIG.targetFile, content, 'utf8');
        log(`成功更新 ${models.length} 条模型数据`);
        
        return true;
    } catch (error) {
        log(`更新数据失败: ${error.message}`);
        return false;
    }
}

// 提取分类标签
function extractCategories(html) {
    const categories = [];
    
    // 提取类似 "文本生成"、"文字转视频" 等分类
    const categoryPattern = /\[(文本生成|文字转视频|文本转图片|图像转视频|Web开发)\]/g;
    
    let match;
    const seen = new Set();
    while ((match = categoryPattern.exec(html)) !== null) {
        if (!seen.has(match[1])) {
            categories.push(match[1]);
            seen.add(match[1]);
        }
    }
    
    return categories;
}

// 更新index.html中的分类标签
function updateCategories(categories) {
    try {
        let content = fs.readFileSync(CONFIG.indexFile, 'utf8');
        
        // 构建新的标签HTML
        let tabsHtml = '';
        categories.forEach((cat, index) => {
            const activeClass = index === 0 ? ' active' : '';
            tabsHtml += `                    <button class="filter-tab${activeClass}">${cat}</button>\n`;
        });
        
        // 替换filter-tabs内容
        const pattern = /(<div class="filter-tabs">)([\s\S]*?)(<\/div>)/;
        const replacement = `$1\n${tabsHtml}                $3`;
        
        content = content.replace(pattern, replacement);
        
        // 写入文件
        fs.writeFileSync(CONFIG.indexFile, content, 'utf8');
        log(`成功更新 ${categories.length} 个分类标签`);
        
        return true;
    } catch (error) {
        log(`更新分类失败: ${error.message}`);
        return false;
    }
}

// 备份数据
function backupData() {
    const timestamp = new Date().toISOString().replace(/:/g, '-').split('.')[0];
    const backupDir = path.join(__dirname, '../backups');
    
    if (!fs.existsSync(backupDir)) {
        fs.mkdirSync(backupDir, { recursive: true });
    }
    
    // 备份model-center.js
    const jsBackup = path.join(backupDir, `model-center-${timestamp}.js`);
    fs.copyFileSync(CONFIG.targetFile, jsBackup);
    
    // 备份index.html
    const htmlBackup = path.join(backupDir, `index-${timestamp}.html`);
    fs.copyFileSync(CONFIG.indexFile, htmlBackup);
    
    log(`数据已备份: ${timestamp}`);
}

// 主函数
async function main() {
    log('========== 开始同步排行榜数据 ==========');
    
    try {
        // 1. 备份当前数据
        log('备份当前数据...');
        backupData();
        
        // 2. 抓取网页内容
        log(`正在抓取数据: ${CONFIG.sourceUrl}`);
        const html = await fetchWebContent(CONFIG.sourceUrl);
        log(`成功获取网页内容 (${html.length} 字节)`);
        
        // 3. 解析排行榜数据
        log('解析排行榜数据...');
        const models = parseLeaderboardData(html);
        
        if (models.length === 0) {
            log('警告: 未能解析到任何模型数据，使用手动数据更新');
            // 这里可以添加备用的手动数据源
            return;
        }
        
        // 4. 提取分类标签
        log('提取分类标签...');
        const categories = extractCategories(html);
        
        // 5. 更新数据文件
        log('更新数据文件...');
        updateModelData(models);
        
        if (categories.length > 0) {
            updateCategories(categories);
        }
        
        log('========== 同步完成 ==========');
        
    } catch (error) {
        log(`同步失败: ${error.message}`);
        log(error.stack);
    }
}

// 如果直接运行此脚本
if (require.main === module) {
    main();
}

module.exports = { main };

File: scripts/sync_leaderboard.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
排行榜数据同步脚本 - Python版本
每天23:00自动从参考网站抓取最新数据并更新
"""

import requests
import json
import re
import os
from datetime import datetime
from pathlib import Path

# 配置
CONFIG = {
    'source_url': 'https://model.aibase.com/zh/leaderboard/text-generation',
    'target_file': Path(__file__).parent.parent / 'js' / 'model-center.js',
    'log_file': Path(__file__).parent.parent / 'logs' / 'sync-log.txt',
    'backup_dir': Path(__file__).parent.parent / 'backups'
}

def log(message):
    """记录日志"""
    timestamp = datetime.now().isoformat()
    log_message = f"[{timestamp}] {message}\n"
    print(log_message.strip())
    
    # 确保logs目录存在
    CONFIG['log_file'].parent.mkdir(parents=True, exist_ok=True)
    
    with open(CONFIG['log_file'], 'a', encoding='utf-8') as f:
        f.write(log_message)

def fetch_web_content(url):
    """抓取网页内容"""
    try:
        headers = {
            'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36'
        }
        response = requests.get(url, headers=headers, timeout=30)
        response.raise_for_status()
        return response.text
    except Exception as e:
        log(f"抓取网页失败: {e}")
        return None

def parse_leaderboard_data(html):
    """解析排行榜数据"""
    models = []
    
    # 使用正则表达式提取数据
    # 根据实际HTML结构调整
    model_pattern = r'#(\d+)\s*\[([^\]]+)\]\([^\)]+\)\s*([^\s]+)\s*([^\s]+)\s*(\d+)'
    
    matches = re.findall(model_pattern, html)
    
    for match in matches:
        models.append({
            'rank': int(match[0]),
            'model': match[1],
            'provider': match[2],
            'license': match[3],
            'score': int(match[4])
        })
    
    return models

def backup_data():
    """备份数据"""
    timestamp = datetime.now().strftime('%Y-%m-%dT%H-%M-%S')
    backup_dir = CONFIG['backup_dir']
    backup_dir.mkdir(parents=True, exist_ok=True)
    
    # 备份model-center.js
    if CONFIG['target_file'].exists():
        js_backup = backup_dir / f"model-center-{timestamp}.js"
        with open(CONFIG['target_file'], 'r', encoding='utf-8') as src:
            with open(js_backup, 'w', encoding='utf-8') as dst:
                dst.write(src.read())
    
    log(f"数据已备份: {timestamp}")

def update_model_data(models):
    """更新model-center.js文件"""
    try:
        with open(CONFIG['target_file'], 'r', encoding='utf-8') as f:
            content = f.read()
        
        # 构建新的数据数组
        new_data_str = json.dumps(models, indent=8, ensure_ascii=False)
        
        # 替换leaderboard数据
        pattern = r'leaderboard:\s*\[[\s\S]*?\]'
        replacement = f'leaderboard: {new_data_str}'
        
        content = re.sub(pattern, replacement, content)
        
        # 写入文件
        with open(CONFIG['target_file'], 'w', encoding='utf-8') as f:
            f.write(content)
        
        log(f"成功更新 {len(models)} 条模型数据")
        return True
        
    except Exception as e:
        log(f"更新数据失败: {e}")
        return False

def main():
    """主函数"""
    log('========== 开始同步排行榜数据 ==========')
    
    try:
        # 1. 备份当前数据
        log('备份当前数据...')
        backup_data()
        
        # 2. 抓取网页内容
        log(f"正在抓取数据: {CONFIG['source_url']}")
        html = fetch_web_content(CONFIG['source_url'])
        
        if not html:
            log('抓取失败，跳过本次同步')
            return
        
        log(f"成功获取网页内容 ({len(html)} 字节)")
        
        # 3. 解析排行榜数据
        log('解析排行榜数据...')
        models = parse_leaderboard_data(html)
        
        if not models:
            log('警告: 未能解析到任何模型数据')
            return
        
        # 4. 更新数据文件
        log('更新数据文件...')
        update_model_data(models)
        
        log('========== 同步完成 ==========')
        
    except Exception as e:
        log(f"同步失败: {e}")
        import traceback
        log(traceback.format_exc())

if __name__ == '__main__':
    main()
